### PR TITLE
Repeat Visitor: Fix threshold reset bug

### LINKED ISSUE #11718: Repeat Visitor Block: Threshold resets when editing post
#### Steps to reproduce the issue

For the Repeat Visitors Block, I’m seeing the counts reset when editing a post again. To reproduce:

    Create a new post and the block.
    Set the count to anything other than 3.
    Publish and view the page.
    Click the edit button on the page to return to the editor.
    See the count has reverted to 3.

#### What I expected
Threshold is maintained.

#### What happened instead
It resets to default.

**Screenshots**
See p8oabR-lg-p2

cc: @a8ck3n

### PR DESCRIPTION
**Pull request by jsnmoon**
Fixes #11718.

#### Changes proposed in this Pull Request:

This casts the threshold attribute value as a number before saving. Previously, it was erroneously being saved as a string.

#### Testing instructions:
0. Spin up a [JN instance](https://jurassic.ninja/create/?gutenpack&branch=fix/repeat-visitor-threshold-persistence).
1. Create a new post and the block.
2. Set the count to anything other than 3.
3. Publish and view the page.
4. Click the edit button on the page to return to the editor.
5. Ensure that your input from (2) has been persisted. 

#### Proposed changelog entry for your changes:

* No changelog necessary.

### CODE REVIEW

**Review by jeherve — DISMISSED** [review]
It seems to work well in my tests. 👍

**Comment by kraftbj** [comment]
Note for merging: Please leave the "Cherry-pick" label for Gutenpack to remove once they've ported this over to the 7.2 Calypso branch to ensure inclusion in 7.2 final.

File: extensions/blocks/repeat-visitor/components/edit.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ class RepeatVisitorEdit extends Component {
 	setCriteria = criteria => this.props.setAttributes( { criteria } );
 	setThreshold = threshold => {
 		if ( /^\d+$/.test( threshold ) && +threshold > 0 ) {
-			this.props.setAttributes( { threshold, isThresholdValid: true } );
+			this.props.setAttributes( { threshold: +threshold, isThresholdValid: true } );
```
Comment (ockham) [inline extensions/blocks/repeat-visitor/components/edit.js]: While we're at it -- can we get rid of the `isThresholdValid` attribute? :grimacing: I find it kinda weird to store an attribute that contains information about the validity of another (`threshold`) -- shouldn't we be able to infer validity of `threshold` from its value at all times?

Comment (bluefuton) [reply extensions/blocks/repeat-visitor/components/edit.js]: Thanks for the suggestion @ockham - makes sense. I'd prefer that we address it in a followup PR after the release if that's alright.

Comment (ockham) [reply extensions/blocks/repeat-visitor/components/edit.js]: The thing is, if we release the block _with_ this attribute and retire it afterwards, we'll need to write deprecations for it, since removing then it will otherwise invalidate our user's existing blocks at that point. If we manage to remove it before the release, we won't have to do that :slightly_smiling_face:

Comment (bluefuton) [reply extensions/blocks/repeat-visitor/components/edit.js]: That's a fair point @ockham - I hadn't considered needing to write a deprecation.

I've had a tinker with the code, and I think we do need to store somewhere that we're currently in an error state. If we get a new threshold set, if it's invalid (like '-5') we don't want to set it in `this.props.attributes.threshold` because then it could be preserved. That's why we've been using a separate attribute (`isThresholdValid`) to keep track of whether the last-entered value was acceptable or not, without storing it.

Would a rename of the attribute help? Is `isError` clearer?

cc @jsnmoon

Comment (simison) [reply extensions/blocks/repeat-visitor/components/edit.js]: @bluefuton note that these attributes can be modified directly in the post content, so checking for validity (or lack of the attribute) each time the value is read is a good idea anyway.

Comment (sirreal) [reply extensions/blocks/repeat-visitor/components/edit.js]: > Would a rename of the attribute help? Is isError clearer?

The concern is that it seems out of place to use attributes for this. Attributes are serialized and stored, so we're storing invalid data, as well as (meta)data that the value is invalid.

I'd suggest that we shouldn't serialize an invalid attribute, and we should be able to compute validity on the fly. That seems to lead to juggling to levels of state: block attributes and local component state.

- The attribute stores the valid value.
- Changes should be pushed to the attribute (setAttribute) there as soon as they're validated. "Dirty" component state (invalid or missing values) should be tracked in local component state. This may mean clearing the attribute when the form is dirty.

In time, it would be nice to land on a solution we can share across components. In Simple Payments, we were exploring Formik to track form state, independent of attributes, but were unable to finish those changes due to time constraints.

Comment (jsnmoon) [reply extensions/blocks/repeat-visitor/components/edit.js]: I was able to replace `this.props.attributes.isThresholdValid` with `this.state.isThresholdValid`. 

It seems to work as expected - let me know what you all think!

Comment (ockham) [reply extensions/blocks/repeat-visitor/components/edit.js]: Thanks a lot @jsnmoon, glad to hear you could replace the attribute with component state! :tada: 

Out of curiosity, with these properties set https://github.com/Automattic/jetpack/blob/cda78ee745f0d6cf8dd85176aa10b4ccbbe18e2b/extensions/blocks/repeat-visitor/components/edit.js#L84-L87 -- how can I actually repro entering an invalid value into that field? :sweat_smile:

File: extensions/blocks/repeat-visitor/components/edit.js [hunk 2]
```diff
@@ -27,17 +27,18 @@ const RADIO_OPTIONS = [
...
-	componentDidMount() {
-		this.props.setAttributes( { isThresholdValid: true } );
-	}
+	state = {
+		isThresholdValid: true,
+	};
 
 	setCriteria = criteria => this.props.setAttributes( { criteria } );
 	setThreshold = threshold => {
 		if ( /^\d+$/.test( threshold ) && +threshold > 0 ) {
```
Comment (ockham) [inline extensions/blocks/repeat-visitor/components/edit.js]: Could we shorten this to
```suggestion
		if ( parsetInt( threshold, 10 ) > 0 ) {
```

(see [MDN](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/parseInt)) (And maybe even use the result from `parsetInt( threshold, 10 )` for `setAttributes`?

Comment (jsnmoon) [reply extensions/blocks/repeat-visitor/components/edit.js]: This approach using `parseInt` is very similar to how our initial validation approach. However, we decided on going with the more rigorous regex check so that we could properly reject improper inputs like `"1.5"` and `"2+1"`.

I think we can safely replace typecasting `threshold` with a more explicit `parseInt` for the `setAttributes` line, however.

What do you think?

File: extensions/blocks/repeat-visitor/components/edit.js (hunk 1 above)
Comment (jsnmoon) [reply extensions/blocks/repeat-visitor/components/edit.js]: @ockham Testers in our call for testing figured out multiple ways :)

You can try inputs like `1+2`, `8eee`, `-14`, and so on.

Comment (ockham) [reply extensions/blocks/repeat-visitor/components/edit.js]: Ah, I see, thanks :slightly_smiling_face: 

Let's remove the `pattern` attr then since that's apparently [not supported](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#Pattern_validation) for this type of input field.

**Comment by simison** [comment]
It seems like the main concerns there are addressed and any other changes would be good for follow up PRs. If we can get this approved and merged soon, we could still make it tonight to Jetpack 7.2 beta 2. :-)

**Comment by jsnmoon** [comment]
@simison Agreed!

File: extensions/blocks/repeat-visitor/components/edit.js (hunk 2 above)
Comment (ockham) [reply extensions/blocks/repeat-visitor/components/edit.js]: Ah, okay. I guess I'm coming from a place of laziness here -- it seems like we're trying to perform a fairly boring task here (non-negative integer validation), so I was hoping we could use really simple tooling for that :slightly_smiling_face: 

I guess it's fine either way. Personally, I like `parseInt` better than the regex and unary `+` (and I don't care as much that it would allow quirky inputs like `"1.5"` and `"2+1"` -- it would transform them into something valid), but I don't care that much really, and I don't want to hold this PR nitpicking over minor things like that forever :sweat_smile:

**Comment by ockham** [comment]
Ah there's one more thing I just noticed, one sec... :sweat_smile:

File: extensions/blocks/repeat-visitor/components/edit.js [hunk 3]
```diff
@@ -77,9 +78,7 @@ class RepeatVisitorEdit extends Component {
 					<TextControl
 						className="wp-block-jetpack-repeat-visitor-threshold"
 						defaultValue={ this.props.attributes.threshold }
```
Comment (ockham) [inline extensions/blocks/repeat-visitor/components/edit.js]: Oh, took me a while to realize we're using `defaultValue` (rather than `value`) here. That's a bit unusual :thinking: What about changing this to a regular `value`? Is it possible that we could drop `state` handling altogether then? (We will probably lose the `Please enter a valid number` message in the process, but users simply wouldn't be able to enter any forbidden chars, which seems good enough -- and pretty standard behavior.)

Comment (jsnmoon) [reply extensions/blocks/repeat-visitor/components/edit.js]: Yes, we could use `value`, drop the `help` property, and completely remove the internal state of the component. However, I think this reduction in code complexity comes at a high UX cost where we silently block user interaction without visible feedback.

The initial implementation for this actually used `value` and we`d received internal feedback during our CFT; I can't imagine how much worse it'd be for the general population.

**Comment by ockham** [comment]
So to wrap up, this is what my suggestion boild down to :grimacing: 

```diff
diff --git a/extensions/blocks/repeat-visitor/components/edit.js b/extensions/blocks/repeat-visitor/components/edit.js
index 3a2e4cd15..791d5a1cb 100644
--- a/extensions/blocks/repeat-visitor/components/edit.js
+++ b/extensions/blocks/repeat-visitor/components/edit.js
@@ -27,18 +27,12 @@ const RADIO_OPTIONS = [
 ];
 
 class RepeatVisitorEdit extends Component {
-       state = {
-               isThresholdValid: true,
-       };
-
        setCriteria = criteria => this.props.setAttributes( { criteria } );
-       setThreshold = threshold => {
-               if ( /^\d+$/.test( threshold ) && +threshold > 0 ) {
-                       this.props.setAttributes( { threshold: +threshold } );
-                       this.setState( { isThresholdValid: true } );
-                       return;
+       setThreshold = value => {
+               const threshold = parseInt( value, 10 );
+               if ( threshold > 0 ) {
+                       this.props.setAttributes( { threshold } );
                }
-               this.setState( { isThresholdValid: false } );
        };
 
        getNoticeLabel() {
@@ -77,12 +71,10 @@ class RepeatVisitorEdit extends Component {
                                >
                                        <TextControl
                                                className="wp-block-jetpack-repeat-visitor-threshold"
-                                               defaultValue={ this.props.attributes.threshold }
-                                               help={ this.state.isThresholdValid ? '' : __( 'Please enter a valid number.' ) }
+                                               value={ this.props.attributes.threshold }
                                                label={ __( 'Visit count threshold' ) }
                                                min="1"
                                                onChange={ this.setThreshold }
-                                               pattern="[0-9]"
                                                type="number"
                                        />
 

```

**Comment by simison** [comment]
>It seems like the main concerns there are addressed and any other changes would be good for follow up PRs. If we can get this approved and merged soon, we could still make it tonight to Jetpack 7.2 beta 2. :-)

To clarify, I meant that all feedback should be discussed and addressed without a feeling of urgency. If it's all in this PR that's great. Let's also make sure that @kraftbj can include some amount of fixes from this PR in the additional beta release going out today — and the rest will be picked still to the final release.

**Review by ockham — APPROVED** [review]
As per discussion in Slack, it was a deliberate decision to go with the present approach, based on testers' feedback during the Call for Testing. I don't want to keep this from merging forever :grimacing: so feel free to :ship:

**Review by kraftbj — APPROVED** [review]

**Comment by sirreal** [comment]
cherry-picked: https://github.com/Automattic/wp-calypso/pull/31793/commits/0e73db8aebae71603b048aa4a17d008c5cf31956
